Source/AIPrototype/Private/Controllers/UnitGroupAIController.cpp

```diff
@@ -0,0 +1,152 @@
+	PerceptionComponent->SetDominantSense(UAISense_Sight::StaticClass());
+}
+
+void AUnitGroupAIController::InitializeControlledUnits(const TArray<AUnitBase*>& Units)
+{
+	for (const auto unit : Units)
+	{
+		m_ControlledUnits.AddUnique(unit);
+	}
+
+	InitBlackboardAlliesNum();
+}
+
+void AUnitGroupAIController::InitBlackboardAlliesNum()
+{
+	GetBlackboardComponent()->SetValueAsInt(CommanderBBKeys::AlliesNum, m_ControlledUnits.Num());
+}
+
+void AUnitGroupAIController::BeginPlay()
+{
+	Super::BeginPlay();
+
+	BindDelegates();
+	RunBehaviorTree(BehaviorTreeAsset);
+}
+
+void AUnitGroupAIController::BindDelegates()
+{
+	PerceptionComponent->OnTargetPerceptionUpdated.AddDynamic(this, &AUnitGroupAIController::OnTargetPerceptionUpdated);
+
+	// consider group movement finished when commander reaches goal location
+	Chain(this, GetPathFollowingComponent()->OnRequestFinished, OnGroupMovementFinishedDelegate, this);
+	OnGroupMovementFinishedDelegate.AddDynamic(this, &AUnitGroupAIController::OnGroupMovementFinished);
+}
+
+void AUnitGroupAIController::OnTargetPerceptionUpdated(AActor* Actor, FAIStimulus Stimulus)
+{
+	if (Stimulus.IsActive())
+	{
+		if (!m_SensedEnemies.Contains(Actor))
+		{
+			m_SensedEnemies.AddUnique(Actor);
+		}
+	}
+	else if (Stimulus.IsExpired())
+	{
+		if (m_SensedEnemies.Contains(Actor))
+		{
+			m_SensedEnemies.Remove(Actor);
+		}
+	}
+
+	UpdateBlackboardEnemiesNum();
+}
+
+
+void AUnitGroupAIController::UpdateBlackboardEnemiesNum()
+{
+	GetBlackboardComponent()->SetValueAsInt(CommanderBBKeys::EnemiesNum, m_SensedEnemies.Num());
+}
+
+void AUnitGroupAIController::Tick(float DeltaSeconds)
+{
+	Super::Tick(DeltaSeconds);
+
+	UpdateBlackboardEnemyLocation();
+}
+
+void AUnitGroupAIController::UpdateBlackboardEnemyLocation()
+{
+	if (m_SensedEnemies.Num() > 0)
+	{
+		GetBlackboardComponent()->SetValueAsVector(CommanderBBKeys::EnemyLocation, CalcSensedEnemiesCenterLocation());
+	}
+	else
+	{
+		GetBlackboardComponent()->ClearValue(CommanderBBKeys::EnemyLocation);
+	}
+}
+
+FVector AUnitGroupAIController::CalcSensedEnemiesCenterLocation() const
+{
+	FVector min_location = FVector(BIG_NUMBER);
+	FVector max_location = FVector(-BIG_NUMBER);
+	FVector enemy_location = FVector::ZeroVector;
+
+	for (const auto enemy_unit : m_SensedEnemies)
+	{
+		if (enemy_unit.IsValid())
+		{
+			enemy_location = enemy_unit->GetActorLocation();
+
+			min_location.X = FMath::Min<float>(enemy_location.X, min_location.X);
+			min_location.Y = FMath::Min<float>(enemy_location.Y, min_location.Y);
+			min_location.Z = FMath::Min<float>(enemy_location.Z, min_location.Z);
+
+			max_location.X = FMath::Max<float>(enemy_location.X, max_location.X);
+			max_location.Y = FMath::Max<float>(enemy_location.Y, max_location.Y);
+			max_location.Z = FMath::Max<float>(enemy_location.Z, max_location.Z);
+		}
+	}
+
+	const auto center_location = (min_location + max_location) * 0.5f;
```

Comment:
> https://docs.unrealengine.com/en-US/API/Runtime/Engine/Kismet/UKismetMathLibrary/GetVectorArrayAverage/index.html

thx 👍

Reply:
Thx, now I know about such method 👍 But it's not the same: my implementation calculates exactly center of **bounding box** of unit group while the method you suggested returns average of vectors. 
I would prefer exactly a center of bounding box because it suits better for a "formation" that units form after spawn - it ends up in a location closer to actual center of spawned units than result of GetVectorArrayAverage method, in case units aren't in square formation.

Reply:
We use average for formations. For example when a player orders to attack an enemy group you want your units to run in a place where most units concentrated. But if use bounding box and one enemy unit get far from formation with other, your units may und up not in a place you want.